pkg/subctl/cmd/logs.go

```diff
@@ -0,0 +1,63 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/submariner-io/submariner-operator/pkg/discovery/network"
+)
+
+func getGatewayPodsDetails(clientSet kubernetes.Interface) {
+	pods, err := network.FindPod(clientSet, "app=submariner-gateway")
+
+	if err != nil {
+		fmt.Println("error getting pods")
+	}
+	for _, pod := range pods.Items {
+		getPodLogs(clientSet, pod.Name)
+	}
+}
+
+func getPodLogs(clientset kubernetes.Interface, podname string) {
+	podLogOpts := corev1.PodLogOptions{}
+	logrequest := clientset.CoreV1().Pods(SubmarinerNamespace).GetLogs(podname, &podLogOpts)
+	logstream, err := logrequest.Stream()
+	if err != nil {
+		fmt.Errorf("error opening log stream %s", err)
+	}
+	defer logstream.Close()
+
+	logs := new(bytes.Buffer)
+	_, err = io.Copy(logs, logstream)
+	if err != nil {
+		fmt.Errorf("error in copy information from logs to buf %s", err)
+	}
+
+	dirname := filepath.Join(strings.Split(podname, "-")[1], "logs")
```

Comment:
Do we need multiple directories? Why not just name files as `connectivity_<podname>.log`

Reply:
my idea to structure it as
```
lighthouse/logs/pod1.log
lighthouse/logs/pod2.log
routeagent/logs/pod1.log
```
and so on for other pods
and similarly for other details
`lighthouse/details/pod1.txt`
and so on

Reply:
I'd prefer to avoid having nested directories. I was thinking something on lines of 

```
gather-<cluster-id>-<timestamp>/
  connectivity-logs-<podname1>.log
  connectivity-logs-<podname2>.log
  service-discovery-logs-<podname1>.log
  broker-resources-gateways.json
```
Avoids unnecessary nested directories and one look at file name you know what the contents are. Instead of submariner or lighthouse, use module names as used in gather command [connectivity, service-discovery etc.]

@mangelajo @skitt @tpantelis @dfarrell07 Appreciate your inputs on this. Let us use this PR to decide on what directory layout and filenames should look like.

Reply:
I think it would be better to put all the logs in one directory. The main reason for collecting the logs would likely be to tar them up and send to support so one directory would probably make that easier. Also for troubleshooting we'd want all the logs atomically captured and stored together. We also may want to allow the user to specify the dir or we generate one with a timestamp.

Reply:
How about this
logs/<cluster-id>-<podname>.log

Reply:
yeah that file name pattern sounds fine. But as described above, we should timestamp the directory - I suggest `submariner-<timestamp>`.

Reply:
> How about this
> `logs/<cluster-id>-<podname>.log`

Don't create separate directory for logs and other resources. We'll be collecting resources in yaml or json format, so log files can be distinguished by `*.log`

Reply:
With the latest push when log collection for all modules is implemented, running `subctl gather` on all clusters might create multiple dirs with different timestamp becasue there is a possibility that the time would have changed mid way and hence some of the logs would be written to new dir.

Here's the proof while running it locally
I ran `subctl gather` and had 2 dirs created

`submariner-20210323231719
submariner-20210323231720`

because the `second` changed.

cluster1 logs got divided in between these 2 dirs

```
ls submariner-20210323231719/
**cluster1-submariner-gateway-qnspm.log**
cluster2-submariner-routeagent-jzrmg.log 
cluster3-submariner-routeagent-8l5sx.log
cluster2-submariner-gateway-jn79f.log     
cluster2-submariner-routeagent-twnmm.log  
cluster3-submariner-routeagent-kl9rr.log
cluster2-submariner-routeagent-552mm.log  
cluster3-submariner-gateway-2mz29.log     
cluster3-submariner-routeagent-mgq9v.log

ls submariner-20210323231720
**cluster1-submariner-routeagent-mtf5h.log  
cluster1-submariner-routeagent-p8fqf.log** 
```

Reply:
The output would be like this

```
$ ./../submariner-operator/bin/subctl gather
Gathering information from cluster2
 ✓ Gathering operator resources
 ⚠ Gather Operator Resources not implemented yet
 ✓ Gathering operator logs
 ✗ Gathering Operator logs not successful error getting logs for pod "submariner-lighthouse-agent-d6bcd4855-gftvq": error opening log stream for pod "submariner-lighthouse-agent-d6bcd4855-gftvq": container "submariner-lighthouse-agent" in pod "submariner-lighthouse-agent-d6bcd4855-gftvq" is waiting to start: image can't be pulled
 ✓ Gathering connectivity logs 
 ✓ Successfully collected Gateway pods logs
 ✓ Successfully collected Routeagent pods logs
 ✓ Gathering connectivity resources
 ⚠ Gather Connectivity Resources not implemented yet
 ✓ Gathering service-discovery logs
 ✗ Gathering Service Discovery logs not successful error getting logs for pod "submariner-lighthouse-agent-d6bcd4855-gftvq": error opening log stream for pod "submariner-lighthouse-agent-d6bcd4855-gftvq": container "submariner-lighthouse-agent" in pod "submariner-lighthouse-agent-d6bcd4855-gftvq" is waiting to start: image can't be pulled
 ✓ Gathering service-discovery resources
 ⚠ Gather ServiceDiscovery Resources not implemented yet
 ✓ Gathering broker logs
 ⚠ No logs to gather on Broker
 ✓ Gathering broker resources
 ⚠ Gather Broker Resources not implemented yet

Gathering information from cluster3
 ✓ Gathering connectivity logs 
 ✓ Successfully collected Gateway pods logs
 ✓ Successfully collected Routeagent pods logs
 ✓ Gathering connectivity resources
 ⚠ Gather Connectivity Resources not implemented yet
 ✓ Gathering service-discovery logs
 ✗ Gathering Service Discovery logs not successful error getting logs for pod "submariner-lighthouse-agent-5ddc466879-flqdn": error opening log stream for pod "submariner-lighthouse-agent-5ddc466879-flqdn": container "submariner-lighthouse-agent" in pod "submariner-lighthouse-agent-5ddc466879-flqdn" is waiting to start: trying and failing to pull image
 ✓ Gathering service-discovery resources
 ⚠ Gather ServiceDiscovery Resources not implemented yet
 ✓ Gathering broker logs
 ⚠ No logs to gather on Broker
 ✓ Gathering broker resources
 ⚠ Gather Broker Resources not implemented yet
 ✓ Gathering operator logs
 ✗ Gathering Operator logs not successful error getting logs for pod "submariner-lighthouse-agent-5ddc466879-flqdn": error opening log stream for pod "submariner-lighthouse-agent-5ddc466879-flqdn": container "submariner-lighthouse-agent" in pod "submariner-lighthouse-agent-5ddc466879-flqdn" is waiting to start: trying and failing to pull image
 ✓ Gathering operator resources
 ⚠ Gather Operator Resources not implemented yet

Gathering information from cluster1
 ✓ Gathering connectivity logs 
 ✓ Successfully collected Gateway pods logs
 ✓ Successfully collected Routeagent pods logs
 ✓ Gathering connectivity resources
 ⚠ Gather Connectivity Resources not implemented yet
 ✓ Gathering service-discovery logs
 ✗ Gathering Service Discovery logs not successful error getting logs for pod "submariner-lighthouse-agent-649c9965db-bvc4m": error opening log stream for pod "submariner-lighthouse-agent-649c9965db-bvc4m": container "submariner-lighthouse-agent" in pod "submariner-lighthouse-agent-649c9965db-bvc4m" is waiting to start: trying and failing to pull image
 ✓ Gathering service-discovery resources
 ⚠ Gather ServiceDiscovery Resources not implemented yet
 ✓ Gathering broker logs
 ⚠ No logs to gather on Broker
 ✓ Gathering broker resources
 ⚠ Gather Broker Resources not implemented yet
 ✓ Gathering operator logs
 ✗ Gathering Operator logs not successful error getting logs for pod "submariner-lighthouse-agent-649c9965db-bvc4m": error opening log stream for pod "submariner-lighthouse-agent-649c9965db-bvc4m": container "submariner-lighthouse-agent" in pod "submariner-lighthouse-agent-649c9965db-bvc4m" is waiting to start: trying and failing to pull image
 ✓ Gathering operator resources
 ⚠ Gather Operator Resources not implemented yet
```

Reply:
The idea is that you put all files from all modules from all clusters under the same timestamped dir which you would create at the beginning of the subclt cmd.